### PR TITLE
Installing Ruby & various lessons: Update to use Ruby 3.1.2

### PR DESCRIPTION
**Pull request by scheals**
Complete the following REQUIRED checkboxes:

-   [x] I have thoroughly read and understand [The Odin Project Contributing Guide](https://github.com/TheOdinProject/theodinproject/blob/main/CONTRIBUTING.md)
-   [x] The title of this PR follows the `location of change: brief description of change` format, e.g. `Intro to HTML and CSS lesson: Fix link text`

Complete the following checkboxes ONLY IF they are applicable to your PR. You can complete them later if they are not currently applicable:
-   [ ] I have previewed all lesson files included in this PR with the [Markdown preview tool](https://www.theodinproject.com/lessons/preview) to ensure the Markdown content is formatted correctly
-   [ ] I have ensured all lesson files included in this PR follow the [Layout Style Guide](https://github.com/TheOdinProject/curriculum/blob/main/LAYOUT_STYLE_GUIDE.md)

<hr>

**1. Because:**

 Ubuntu 22.04 comes with openSSL 3.0 installed and only Ruby 3.1+ started being compatible with it. Because of this, installing Ruby 3.0.3 is not possible on 22.04 without tinkering with openSSL installation which is not recommended. 20.04 installs are unaffected.
 
Part of #24112

**2. This PR:**

* Updates the installation instructions to use Ruby 3.1.2 
* Updates various links to the documentation to point to the 3.1.2 documentation
* Updates various links to the rubyapi version of documentation to point to the 3.1 documentation

**3. Additional Information:**

I also have to change various `.ruby-version` files to be 3.1.2, update the gems & the bundler in the repos and change description in ruby_testing, like this issue's PRs: #23434

### CODE REVIEW

**Comment by ChargrilledChook** [comment]
Thank you for taking the time to work on this @scheals. I'd like to check with the team if there are any issues with switching over at the moment, assuming no issues there I aim to review all of your changes over the next couple of days.

File: ruby/files_and_serialization/project_event_manager.md [hunk 1]
```diff
@@ -125,7 +125,7 @@ The first few rows of the CSV file you downloaded look like this:
 
 #### Read the File Contents
 
-[File](http://rubydoc.info/stdlib/core/File) is a core ruby class that allows
+[File](http://rubydoc.info/stdlib/core-3.1.2/File) is a core-3.1.2 ruby class that allows
```
Comment (ChargrilledChook) [inline ruby/files_and_serialization/project_event_manager.md:128]: @scheals I think your find and replace was a little over zealous here :wink: 

```suggestion
[File](http://rubydoc.info/stdlib/core-3.1.2/File) is a core ruby class that allows
```

File: ruby/files_and_serialization/project_event_manager.md [hunk 2]
```diff
@@ -4,12 +4,12 @@ Please note this tutorial has been adapted from The Turing School's and Jump Sta
 
 After completing this tutorial, you will be able to:
 
-* manipulate [file](http://rubydoc.info/stdlib/core/File) input and output
+* manipulate [file](http://rubydoc.info/stdlib/core-3.1.2/File) input and output
```
Comment (ChargrilledChook) [inline ruby/files_and_serialization/project_event_manager.md:7]: This link doesn't work for me, I think version agnostic links should generally be fine to leave

File: ruby/files_and_serialization/project_event_manager.md [hunk 3]
```diff
@@ -4,12 +4,12 @@ Please note this tutorial has been adapted from The Turing School's and Jump Sta
...
 * read content from a [CSV](http://rubydoc.info/stdlib/csv/file/README.rdoc) (Comma Separated Value) file
 * transform it into a standardized format
 * utilize the data to contact a remote service
 * populate a template with user data
-* manipulate [strings](http://rubydoc.info/stdlib/core/String)
+* manipulate [strings](http://rubydoc.info/stdlib/core-3.1.2/String)
```
Comment (ChargrilledChook) [inline ruby/files_and_serialization/project_event_manager.md:12]: This one too

File: ruby/files_and_serialization/project_event_manager.md [hunk 4]
```diff
@@ -149,7 +149,7 @@ filepath on the filesystem through `File.exist? "event_attendees.csv"`.
 
 Reading and displaying the entire contents of the file showed us how to quickly
 access the data. Our goal is to display the first names of all the attendees.
-There are numerous [String](http://rubydoc.info/stdlib/core/String) methods
+There are numerous [String](http://rubydoc.info/stdlib/core-3.1.2/String) methods
```
Comment (ChargrilledChook) [inline ruby/files_and_serialization/project_event_manager.md:152]: Broken link here

File: ruby/files_and_serialization/project_event_manager.md [hunk 5]
```diff
@@ -203,7 +203,7 @@ array of columns. The separation of the columns can be identified by the comma
 `","` separator. We want to split the string into pieces wherever we see a comma.
 
 Ruby's
-[String#split](http://rubydoc.info/stdlib/core/String#split-instance_method)
+[String#split](http://rubydoc.info/stdlib/core-3.1.2/String#split-instance_method)
```
Comment (ChargrilledChook) [inline ruby/files_and_serialization/project_event_manager.md:206]: Broken link here

File: ruby/files_and_serialization/project_event_manager.md [hunk 6]
```diff
@@ -286,7 +286,7 @@ end
 ~~~
 
 This is a such a common operation that Array defines
-[Array#each_with_index](http://rubydoc.info/stdlib/core/Enumerable#each_with_index-instance_method).
+[Array#each_with_index](http://rubydoc.info/stdlib/core-3.1.2/Enumerable#each_with_index-instance_method).
```
Comment (ChargrilledChook) [inline ruby/files_and_serialization/project_event_manager.md:289]: Broken link here

File: ruby/files_and_serialization/project_event_manager.md [hunk 7]
```diff
@@ -339,7 +339,7 @@ We will then use this new parser to access our attendees' zip codes.
 
 #### Switching over to use the CSV Library
 
-Ruby's core language comes with a wealth of great classes. Not all of them are
+Ruby's core-3.1.2 language comes with a wealth of great classes. Not all of them are
```
Comment (ChargrilledChook) [inline ruby/files_and_serialization/project_event_manager.md:342]: Extra version here

File: ruby/files_and_serialization/project_event_manager.md [hunk 8]
```diff
@@ -504,15 +504,15 @@ paths:
   * Use a `while` or `until` loop to prepend zeros until the length is five
   * Calculate the length of the current zip code and add missing zeros to the front
   * Add five zeros to the front of the current zip code and then trim the last five digits
-  * Use [String#rjust](http://rubydoc.info/stdlib/core/String#rjust-instance_method) to append zeros to the front of the string.
+  * Use [String#rjust](http://rubydoc.info/stdlib/core-3.1.2/String#rjust-instance_method) to append zeros to the front of the string.
```
Comment (ChargrilledChook) [inline ruby/files_and_serialization/project_event_manager.md:507]: Broken link here

File: ruby/files_and_serialization/project_event_manager.md [hunk 9]
```diff
@@ -504,15 +504,15 @@ paths:
...
 
 #### Handling Bad and Good Zip Codes
 
 The following solution employs:
 
-* [String#length](http://rubydoc.info/stdlib/core/String#length-instance_method) - returns the length of the string.
-* [String#rjust](http://rubydoc.info/stdlib/core/String#rjust-instance_method) - to pad the string with zeros.
-* [String#slice](http://rubydoc.info/stdlib/core/String#slice-instance_method) - to create sub-strings either through the `slice` method or the array-like notation `[]`
+* [String#length](http://rubydoc.info/stdlib/core-3.1.2/String#length-instance_method) - returns the length of the string.
```
Comment (ChargrilledChook) [inline ruby/files_and_serialization/project_event_manager.md:513]: Broken links here

File: ruby/files_and_serialization/project_event_manager.md [hunk 10]
```diff
@@ -681,15 +681,15 @@ A good rule when developing in Ruby is to favor coercing values into similar
 values so that they will behave the same. We have a special case to deal
 specifically with a `nil` value. It would be much easier if instead of checking
 for a nil value, we convert the `nil` into a string with
-[NilClass#to_s](http://rubydoc.info/stdlib/core/NilClass#to_s-instance_method).
+[NilClass#to_s](http://rubydoc.info/stdlib/core-3.1.2/NilClass#to_s-instance_method).
```
Comment (ChargrilledChook) [inline ruby/files_and_serialization/project_event_manager.md:684]: Broken link here

File: ruby/files_and_serialization/project_event_manager.md [hunk 11]
```diff
@@ -681,15 +681,15 @@ A good rule when developing in Ruby is to favor coercing values into similar
...
 
 ~~~ruby
 $ nil.to_s
 => ""
 ~~~
 
 Examining
-[String#rjust](http://rubydoc.info/stdlib/core/String#rjust-instance_method) in
+[String#rjust](http://rubydoc.info/stdlib/core-3.1.2/String#rjust-instance_method) in
```
Comment (ChargrilledChook) [inline ruby/files_and_serialization/project_event_manager.md:692]: Broken link here

File: ruby/files_and_serialization/project_event_manager.md [hunk 12]
```diff
@@ -700,7 +700,7 @@ $ "123456".rjust(5, '0')
 ~~~
 
 Lastly, examining
-[String#slice](http://rubydoc.info/stdlib/core/String#slice-instance_method) in
+[String#slice](http://rubydoc.info/stdlib/core-3.1.2/String#slice-instance_method) in
```
Comment (ChargrilledChook) [inline ruby/files_and_serialization/project_event_manager.md:703]: Broken link here

File: ruby/files_and_serialization/project_event_manager.md [hunk 13]
```diff
@@ -842,10 +842,10 @@ Running our application, we find an error.
 
 ~~~bash
 $ ruby lib/event_manager.rb
-/ruby-2.4.0/gems/google-api-client-0.15.0/lib/google/apis/core/http_command.rb:218:in 'check_status': parseError: Failed to parse address (Google::Apis::ClientError)
+/ruby-2.4.0/gems/google-api-client-0.15.0/lib/google/apis/core-3.1.2/http_command.rb:218:in 'check_status': parseError: Failed to parse address (Google::Apis::ClientError)
```
Comment (ChargrilledChook) [inline ruby/files_and_serialization/project_event_manager.md:845]: Broken link here

File: ruby/files_and_serialization/project_event_manager.md [hunk 14]
```diff
@@ -930,9 +930,9 @@ Sarah 33703 ["Marco Rubio", "Bill Nelson", "C. Young"]
 
 The problem now is that when using string interpolation, Ruby is converting our new array of legislator names into a string, but Ruby does not know exactly how _you_ want to display the contents.
 
-We need to explicitly convert our array of legislator names to a string. This way we are sure it will output correctly. This could be tedious work except Array again comes to the rescue with the [Array#join](http://rubydoc.info/stdlib/core/Array#join-instance_method) method.
+We need to explicitly convert our array of legislator names to a string. This way we are sure it will output correctly. This could be tedious work except Array again comes to the rescue with the [Array#join](http://rubydoc.info/stdlib/core-3.1.2/Array#join-instance_method) method.
```
Comment (ChargrilledChook) [inline ruby/files_and_serialization/project_event_manager.md:933]: Broken link here

File: ruby/files_and_serialization/project_event_manager.md [hunk 15]
```diff
@@ -1134,8 +1134,8 @@ For each of our attendees we want to replace the `FIRST_NAME` and `LEGISLATORS`
 * We need to find all instances of `LEGISLATORS` and replace them with the individual's representatives.
 
 Our template is a String of text which has two methods for replacing text:
-[String#gsub](http://rubydoc.info/stdlib/core/String#gsub-instance_method) and
-[String#gsub!](http://rubydoc.info/stdlib/core/String#gsub%21-instance_method).
+[String#gsub](http://rubydoc.info/stdlib/core-3.1.2/String#gsub-instance_method) and
```
Comment (ChargrilledChook) [inline ruby/files_and_serialization/project_event_manager.md:1137]: A few more in this file, I'll stop here pointing out each one individually :laughing:

**Review by ChargrilledChook — CHANGES_REQUESTED** [review]
Thanks very much for your work on this!

It's mostly fine, but there are some issues where you've updated links from version agnostic urls to versioned ones, which has broken them. 

There's also a handful of times where you've updated `core` in plain text to have a version next to it too.

If you can fix these up I think it should be ready to merge :+1:

**Comment by scheals** [comment]
> Thanks very much for your work on this!
> 
> It's mostly fine, but there are some issues where you've updated links from version agnostic urls to versioned ones, which has broken them.
> 
> There's also a handful of times where you've updated `core` in plain text to have a version next to it too.
> 
> If you can fix these up I think it should be ready to merge +1

Ahhh! I didn't pick up on the fact that this lesson used `rubydoc.info` and not `ruby-doc.info`... I would prefer to change the links to the original `ruby-doc.info` here, then. I like `rubyapi.org` the most but I guess `ruby-doc.info` is fine too.

I made it version-specific so it is more in-line with the rest of curriculum which points to particular version.

And sorry for the overzealous core overwrites! There was some back-and-forth with my find-and-replace that probably resulted in this.

**Comment by ChargrilledChook** [comment]
> Ahhh! I didn't pick up on the fact that this lesson used `rubydoc.info` and not `ruby-doc.info`... I would prefer to change the links to the original `ruby-doc.info` here, then. I like `rubyapi.org` the most but I guess `ruby-doc.info` is fine too.
No worries at all, it's a fairly subtle distinction.

IMO all 3 have a place; `ruby-doc` being the most 'official', `rubydoc` being the most extensive (it's auto-generated and also covers gems) and `rubyapi` being the best formatted (albeit more limited in content than the other two). 

I think `rubyapi` is my preference too, especially for core classes, but for the sake of this particular set of changes, I think it would be better to update the links 'as they are' for now. 

Certainly open to discussing switching over / being a bit more consistent in how we recommend docs, though

> I made it version-specific so it is more in-line with the rest of curriculum which points to particular version.
Yeah, the logic makes sense to me :+1: although having agnostic links is probably better to avoid having to do these rewrites every time we update

> And sorry for the overzealous core overwrites! There was some back-and-forth with my find-and-replace that probably resulted in this.
All good :smiley:

**Comment by scheals** [comment]
> IMO all 3 have a place; `ruby-doc` being the most 'official', `rubydoc` being the most extensive (it's auto-generated and also covers gems) and `rubyapi` being the best formatted (albeit more limited in content than the other two).
> 
> I think `rubyapi` is my preference too, especially for core classes, but for the sake of this particular set of changes, I think it would be better to update the links 'as they are' for now.
> 

Alright, I'll revert the changes then! I do know how to convert rubydoc links to ruby-doc links now, though! :laughing: 

> Certainly open to discussing switching over / being a bit more consistent in how we recommend docs, though
> 

I believe that having a "more close encounter" with rubyapi might be beneficial. It's still the case that people in the ruby channels don't know about it and honestly it's a great resource. But that's for another issue.

> > I made it version-specific so it is more in-line with the rest of curriculum which points to particular version.
> 
> Yeah, the logic makes sense to me +1 although having agnostic links is probably better to avoid having to do these rewrites every time we update

I followed what's been done before in the curriculum at large, I think rubydoc currently links to 3.0.3 docs with its agnostic links also. Not sure whether converting everything to be agnostic makes sense because we don't know how up-to-date the curriculum is going to be in the future. The 3.0.3 -> 3.1.2 change has been pretty rapid.

Oh also rubydoc links to some [rdoc](https://rubydoc.info/stdlib/csv/file/README.rdoc) from possibly 2005 that I wasn't able to track down on ruby-doc anyway.

**Comment by scheals** [comment]
@ChargrilledChook Turns out Event Manager lesson features rubydoc, ruby-doc and rubyapi links. Got my find-&-replace overzealous again. As such, I made two ruby-doc links version agnostic which seems more in line with rubydoc being agnostic too, but now rubyapi is the only version-specific one. Not sure whether rubyapi has an agnostic URL, it seems like you can chop the version off and it'll still navigate things but not sure how dependable that is.

**Review by ChargrilledChook — COMMENTED** [review]
LGTM :+1:

**Review by ChargrilledChook — APPROVED** [review]

**Pull request merged**